src/main/java/org/apache/mesos/scheduler/DefaultScheduler.java

```diff
@@ -0,0 +1,309 @@
+            logger.info(String.format("- Offer %d: %s", i + 1, TextFormat.shortDebugString(offers.get(i))));
+        }
+    }
+
+    private void declineOffers(SchedulerDriver driver, List<Protos.OfferID> acceptedOffers, List<Protos.Offer> offers) {
+        for (Protos.Offer offer : offers) {
+            Protos.OfferID offerId = offer.getId();
+            if (!acceptedOffers.contains(offerId)) {
+                logger.info("Declining offer: " + offerId.getValue());
+                driver.declineOffer(offerId);
+            }
+        }
+    }
+
+    private ResourceCleanerScheduler getCleanerScheduler() {
+        try {
+            ResourceCleaner cleaner = new ResourceCleaner(stateStore.getExpectedResources());
+            return new ResourceCleanerScheduler(cleaner, offerAccepter);
+        } catch (Exception ex) {
+            logger.error("Failed to construct ResourceCleaner", ex);
+            return null;
+        }
+    }
+
+    private List<Protos.Offer> filterAcceptedOffers(List<Protos.Offer> offers, List<Protos.OfferID> acceptedOfferIds) {
+        List<Protos.Offer> filteredOffers = new ArrayList<>();
+
+        for (Protos.Offer offer : offers) {
+            if (!offerAccepted(offer, acceptedOfferIds)) {
+                filteredOffers.add(offer);
+            }
+        }
+
+        return filteredOffers;
+    }
+
+    private boolean offerAccepted(Protos.Offer offer, List<Protos.OfferID> acceptedOfferIds) {
+        for (Protos.OfferID acceptedOfferId: acceptedOfferIds) {
+            if (acceptedOfferId.equals(offer.getId())) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    @SuppressWarnings({"DM_EXIT"})
+    private void hardExit(SchedulerErrorCode errorCode) {
+        System.exit(errorCode.ordinal());
+    }
+
+    @Override
+    public void registered(SchedulerDriver driver, Protos.FrameworkID frameworkId, Protos.MasterInfo masterInfo) {
+        logger.info("Registered framework with frameworkId: " + frameworkId.getValue());
+        initialize();
+
+        try {
+            stateStore.storeFrameworkId(frameworkId);
+        } catch (Exception e) {
+            logger.error(String.format(
+                    "Unable to store registered framework ID '%s'", frameworkId.getValue()), e);
+            hardExit(SchedulerErrorCode.REGISTRATION_FAILURE);
+        }
+    }
+
+    @Override
+    public void reregistered(SchedulerDriver driver, Protos.MasterInfo masterInfo) {
+        logger.error("Re-registration implies we were unregistered.");
+        hardExit(SchedulerErrorCode.RE_REGISTRATION);
+    }
+
+    @Override
+    public void resourceOffers(SchedulerDriver driver, List<Protos.Offer> offers) {
+        executor.execute(new Runnable() {
+            @Override
+            public void run() {
+                logOffers(offers);
+
+                // Task Reconciliation
+                reconciler.reconcile(driver);
```

Comment:
In this implementation, we don't reconcile right after we are registered OR re-registered. Instead we wait for the first offer cycle. This is sub-optimal. Is there any specific reason why it's done this way?

In busy clusters the scheduler will waste time until first offer cycle sitting idle.